src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/CaCaseloadService.kt

```diff
@@ -0,0 +1,492 @@
+      return@map CaCase(
+        kind = licence?.kind,
+        licenceId = licence?.licenceId,
+        licenceVersionOf = licence?.versionOf,
+        name = licence.let { "${it?.forename} ${it?.surname}" },
+        prisonerNumber = licence?.nomisId!!,
+        releaseDate = releaseDate,
+        releaseDateLabel = if (licence.actualReleaseDate != null) {
+          "Confirmed release date"
+        } else {
+          "CRD"
+        },
+        licenceStatus = licence.licenceStatus,
+        nomisLegalStatus = p.prisoner.legalStatus,
+        lastWorkedOnBy = licence.updatedByFullName,
+        isDueForEarlyRelease = licence.isDueForEarlyRelease,
+        isInHardStopPeriod = licence.isInHardStopPeriod,
+        tabType = determineCaViewCasesTab(
+          p.prisoner,
+          p.cvl,
+          licence,
+        ),
+        probationPractitioner = ProbationPractitioner(staffUsername = licence.comUsername),
+      )
+    }.filterNotNull()
+  }
+
+  private fun determineCaViewCasesTab(
+    nomisRecord: Prisoner,
+    cvlFields: CvlFields,
+    licence: LicenceSummary?,
+  ): CaViewCasesTab {
+    val releaseDate = (licence?.actualReleaseDate ?: licence?.conditionalReleaseDate) ?: selectReleaseDate(
+      nomisRecord,
+    )
+
+    if (licence != null &&
+      isAttentionNeeded(
+        licence.licenceStatus,
+        licence.licenceStartDate,
+        releaseDate,
+      )
+    ) {
+      return CaViewCasesTab.ATTENTION_NEEDED
+    }
+    val isDueToBeReleasedInTheNextTwoWorkingDays =
+      licence?.isDueToBeReleasedInTheNextTwoWorkingDays ?: cvlFields.isDueToBeReleasedInTheNextTwoWorkingDays
+
+    return when {
+      isDueToBeReleasedInTheNextTwoWorkingDays -> CaViewCasesTab.RELEASES_IN_NEXT_TWO_WORKING_DAYS
+      else -> CaViewCasesTab.FUTURE_RELEASES
+    }
+  }
+
+  private fun isAttentionNeeded(
+    status: LicenceStatus,
+    licenceStartDate: LocalDate?,
+    releaseDate: LocalDate?,
+    overrideClock: Clock? = null,
+  ): Boolean {
+    val now = overrideClock ?: clock
+    val today = LocalDate.now(now)
+
+    val noReleaseDates = releaseDate == null
+
+    val missingDates = listOf(
+      IN_PROGRESS,
+      SUBMITTED,
+      APPROVED,
+      NOT_STARTED,
+    ).contains(status) &&
+      noReleaseDates
+    val startDateInPast = licenceStartDate != null && status == APPROVED && licenceStartDate.isBefore(today)
+
+    return missingDates || startDateInPast
+  }
+
+  private fun selectReleaseDate(nomisRecord: Prisoner): LocalDate? {
+    return nomisRecord.confirmedReleaseDate ?: nomisRecord.conditionalReleaseDate
+  }
+
+  private fun getPrisonersApproachingRelease(
+    prisonCaseload: List<String>?,
+    overrideClock: Clock? = null,
+  ): Page<CaseloadItem>? {
+    val now = overrideClock ?: clock
+    val weeksToAdd: Long = 4
+    val today = LocalDate.now(now)
+    val todayPlusFourWeeks = LocalDate.now(now).plusWeeks(weeksToAdd)
+    return prisonCaseload?.toSet()?.let {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Why does this need to take the prison codes as a set?

Reply:
Because of **prisonerSearchApiClient.searchPrisonersByReleaseDate** parma **prisonIds: Set<String>**.